File: 2025/day10/clojure/solution.clj
#!/usr/bin/env clojure

(ns day10.solution
  (:require [clojure.string :as str]))

;; ===== PARSING =====

(defn parse-line-part1 [line]
  "Parse a machine line for Part 1 - extract indicator pattern and buttons."
  (let [indicator-match (re-find #"\[([.#]+)\]" line)
        indicator (second indicator-match)
        n-lights (count indicator)

        ;; Target state: 1 where # appears
        target (reduce (fn [acc [i c]]
                         (if (= c \#)
                           (bit-or acc (bit-shift-left 1 i))
                           acc))
                       0
                       (map-indexed vector indicator))

        ;; Extract button schematics
        buttons (mapv (fn [match]
                        (let [indices (mapv #(Integer/parseInt %)
                                          (str/split (second match) #","))]
                          (reduce (fn [mask idx]
                                    (bit-or mask (bit-shift-left 1 idx)))
                                  0
                                  indices)))
                      (re-seq #"\(([0-9,]+)\)" line))]
    [n-lights target buttons]))

(defn parse-line-part2 [line]
  "Parse a machine line for Part 2 - extract joltage requirements and buttons."
  (let [joltage-match (re-find #"\{([0-9,]+)\}" line)
        joltage (mapv #(Integer/parseInt %)
                     (str/split (second joltage-match) #","))
        n-counters (count joltage)

        ;; Extract button schematics as lists of indices
        buttons (mapv (fn [match]
                        (mapv #(Integer/parseInt %)
                             (str/split (second match) #",")))
                      (re-seq #"\(([0-9,]+)\)" line))]
    [n-counters joltage buttons]))

;; ===== PART 1: BRUTE FORCE XOR =====

(defn count-bits [n]
  "Count number of 1 bits in n."
  (loop [x n cnt 0]
    (if (zero? x)
      cnt
      (recur (bit-shift-right x 1)
             (+ cnt (bit-and x 1))))))

(defn solve-machine-part1 [n-lights target buttons]
  "Brute force: try all 2^n button combinations."
  (let [n-buttons (count buttons)]
    (loop [mask 0 min-presses Long/MAX_VALUE]
      (if (>= mask (bit-shift-left 1 n-buttons))
        (if (= min-presses Long/MAX_VALUE) 0 min-presses)
        (let [state (reduce (fn [s i]
                              (if (bit-test mask i)
                                (bit-xor s (buttons i))
                                s))
                            0
                            (range n-buttons))
              presses (count-bits mask)]
          (recur (inc mask)
                 (if (= state target)
                   (min min-presses presses)
                   min-presses)))))))

;; ===== PART 2: GAUSSIAN ELIMINATION WITH RATIOS =====

(defn mapv-indexed [f coll]
  "Like mapv but with index."
  (vec (map-indexed f coll)))

(defn is-nonneg-int? [x]
  "Check if x is a non-negative integer (works with Ratio and BigInt)."
  (and (>= x 0)
       (or (integer? x)
           (and (ratio? x) (= 1 (denominator x))))))

(defn to-int [x]
  "Convert rational to integer (works with Ratio and BigInt)."
  (if (ratio? x)
    (numerator x)
    x))

(defn gaussian-elimination-rref [A b]
  "Gaussian elimination to RREF. Returns augmented matrix in RREF.
   A is n-rows x n-cols matrix, b is n-rows vector.
   Uses Clojure's Ratio type for exact rational arithmetic."
  (let [n-rows (count A)
        n-cols (count (first A))
        ;; Create augmented matrix [A | b] with ratios
        aug (vec (map-indexed (fn [i row]
                                (vec (concat (map #(rationalize %) row)
                                           [(rationalize (b i))])))
                              A))]
    (loop [pivot-row 0 col 0 aug aug pivot-cols []]
      (if (or (>= pivot-row n-rows) (>= col n-cols))
        {:aug aug :pivot-cols pivot-cols}

        ;; Find non-zero entry in this column
        (let [found (first (filter #(not= 0 (get-in aug [% col]))
                                  (range pivot-row n-rows)))]
          (if (nil? found)
            ;; No pivot, move to next column
            (recur pivot-row (inc col) aug pivot-cols)

            ;; Swap rows
            (let [aug (if (= found pivot-row)
                       aug
                       (assoc aug pivot-row (aug found)
                                  found (aug pivot-row)))

                  ;; Scale pivot row
                  scale (get-in aug [pivot-row col])
                  aug (update aug pivot-row
                             (fn [row]
                               (mapv #(/ % scale) row)))

                  ;; Eliminate column in all other rows
                  aug (reduce (fn [aug row]
                               (if (or (= row pivot-row)
                                      (= 0 (get-in aug [row col])))
                                 aug
                                 (let [factor (get-in aug [row col])]
                                   (update aug row
                                          (fn [r]
                                            (mapv-indexed
                                             (fn [c v]
                                               (- v (* factor (get-in aug [pivot-row c]))))
                                             r))))))
                             aug
                             (range n-rows))]
              (recur (inc pivot-row) (inc col) aug
                     (conj pivot-cols [col pivot-row])))))))))

(defn solve-machine-part2 [n-counters joltage buttons]
  "Solve Part 2 using Gaussian elimination and null space search."
  (let [n-buttons (count buttons)]
    (if (zero? n-buttons)
      (if (every? zero? joltage) 0 Long/MAX_VALUE)

      ;; Build matrix A (n_counters x n_buttons)
      (let [A (vec (for [i (range n-counters)]
                     (vec (for [j (range n-buttons)]
                            (if (some #(= % i) (buttons j)) 1 0)))))
            b (vec joltage)

            ;; Gaussian elimination
            {:keys [aug pivot-cols]} (gaussian-elimination-rref A b)

            ;; Check for inconsistency
            inconsistent? (some (fn [row-idx]
                                 (let [row (aug row-idx)]
                                   (and (every? zero? (take n-buttons row))
                                        (not= 0 (last row)))))
                               (range (count pivot-cols) n-counters))]

        (if inconsistent?
          Long/MAX_VALUE

          ;; Identify free variables
          (let [pivot-col-set (set (map first pivot-cols))
                free-vars (vec (filter #(not (pivot-col-set %))
                                      (range n-buttons)))]

            (if (empty? free-vars)
              ;; Unique solution
              (let [solution (vec (repeat n-buttons 0N))
                    solution (reduce (fn [sol [col row]]
                                      (assoc sol col (last (aug row))))
                                    solution
                                    pivot-cols)]
                (if (every? is-nonneg-int? solution)
                  (reduce + (map to-int solution))
                  Long/MAX_VALUE))

              ;; With free variables, search for optimal solution
              (let [;; Extract null space vectors
                    null-vectors (mapv (fn [fv]
                                        (let [vec (vec (repeat n-buttons 0N))
                                              vec (assoc vec fv 1N)]
                                          (reduce (fn [v [col row]]
                                                   (assoc v col (- (get-in aug [row fv]))))
                                                 vec
                                                 pivot-cols)))
                                      free-vars)

                    ;; Extract particular solution
                    particular (reduce (fn [sol [col row]]
                                        (assoc sol col (last (aug row))))
                                      (vec (repeat n-buttons 0N))
                                      pivot-cols)

                    ;; Search for optimal solution
                    max-j (apply max joltage)
                    n-free (count free-vars)]

                (cond
                  ;; 1D search
                  (= n-free 1)
                  (let [nv (first null-vectors)
                        ;; Compute bounds
                        [t-low t-high]
                        (reduce (fn [[low high] j]
                                 (let [p (particular j)
                                       nv-j (nv j)]
                                   (cond
                                     (zero? nv-j) (if (< p 0)
                                                   [Double/POSITIVE_INFINITY Double/NEGATIVE_INFINITY]
                                                   [low high])
                                     (pos? nv-j) [(max low (/ (- p) nv-j)) high]
                                     :else [low (min high (/ (- p) nv-j))])))
                               [Double/NEGATIVE_INFINITY Double/POSITIVE_INFINITY]
                               (range n-buttons))

                        t-low-int (long (Math/ceil (double t-low)))
                        t-high-int (long (Math/floor (double t-high)))]

                    (loop [t t-low-int min-total Long/MAX_VALUE]
                      (if (> t t-high-int)
                        (if (= min-total Long/MAX_VALUE) 0 min-total)
                        (let [solution (mapv (fn [j]
                                              (+ (particular j) (* t (get nv j))))
                                            (range n-buttons))
                              valid? (every? is-nonneg-int? solution)]
                          (recur (inc t)
                                 (if valid?
                                   (min min-total (reduce + (map to-int solution)))
                                   min-total))))))

                  ;; 2D search
                  (= n-free 2)
                  (let [nv0 (nth null-vectors 0)
                        nv1 (nth null-vectors 1)
                        bound max-j]
                    (loop [t0 (- bound) min-total Long/MAX_VALUE]
                      (if (> t0 bound)
                        (if (= min-total Long/MAX_VALUE) 0 min-total)
                        (let [inter (mapv (fn [j]
                                           (+ (particular j) (* t0 (nv0 j))))
                                         (range n-buttons))

                              ;; Compute bounds for t1
                              [t1-low t1-high]
                              (reduce (fn [[low high] j]
                                       (let [p (inter j)
                                             nv-j (nv1 j)]
                                         (cond
                                           (zero? nv-j) [low high]
                                           (pos? nv-j) [(max low (/ (- p) nv-j)) high]
                                           :else [low (min high (/ (- p) nv-j))])))
                                     [Double/NEGATIVE_INFINITY Double/POSITIVE_INFINITY]
                                     (range n-buttons))

                              t1-low-int (long (Math/ceil (double t1-low)))
                              t1-high-int (long (Math/floor (double t1-high)))]

                          (recur (inc t0)
                                 (loop [t1 t1-low-int min-t min-total]
                                   (if (> t1 t1-high-int)
                                     min-t
                                     (let [solution (mapv (fn [j]
                                                           (+ (inter j) (* t1 (nv1 j))))
                                                         (range n-buttons))
                                           valid? (every? is-nonneg-int? solution)]
                                       (recur (inc t1)
                                              (if valid?
                                                (min min-t (reduce + (map to-int solution)))
                                                min-t))))))))))

                  ;; 3+D: use recursive search with dynamic bounds
                  (<= n-free 6)
                  (let [bound (* 2 max-j)]
                    (letfn [(search [idx partial min-total]
                              (if (= idx n-free)
                                ;; Evaluate solution
                                (if (every? is-nonneg-int? partial)
                                  (min min-total (reduce + (map to-int partial)))
                                  min-total)

                                ;; Compute bounds for current free var
                                (let [nv (nth null-vectors idx)
                                      [t-low t-high]
                                      (reduce (fn [[low high] j]
                                               (let [p (partial j)
                                                     nv-j (nv j)]
                                                 (cond
                                                   (zero? nv-j) [low high]
                                                   (pos? nv-j) [(max low (/ (- p) nv-j)) high]
                                                   :else [low (min high (/ (- p) nv-j))])))
                                             [Double/NEGATIVE_INFINITY Double/POSITIVE_INFINITY]
                                             (range n-buttons))

                                      t-low-int (max (long (Math/ceil (double t-low)))
                                                    (- bound))
                                      t-high-int (min (long (Math/floor (double t-high)))
                                                     bound)]

                                  (loop [t t-low-int min-t min-total]
                                    (if (> t t-high-int)
                                      min-t
                                      (let [new-partial (mapv (fn [j]
                                                               (+ (partial j) (* t (nv j))))
                                                             (range n-buttons))]
                                        (recur (inc t)
                                               (search (inc idx) new-partial min-t))))))))]
                      (let [result (search 0 particular Long/MAX_VALUE)]
                        (if (= result Long/MAX_VALUE) 0 result))))

                  :else 0))))))))))

;; ===== MAIN =====

(defn part1 [lines]
  "Find minimum total button presses for all machines."
  (reduce (fn [total line]
            (if (str/blank? line)
              total
              (let [[n-lights target buttons] (parse-line-part1 line)]
                (+ total (solve-machine-part1 n-lights target buttons)))))
          0
          lines))

(defn part2 [lines]
  "Find minimum total button presses for joltage configuration."
  (reduce (fn [total line]
            (if (str/blank? line)
              total
              (let [[n-counters joltage buttons] (parse-line-part2 line)]
                (+ total (solve-machine-part2 n-counters joltage buttons)))))
          0
          lines))

(defn -main []
  (let [input (slurp "../input.txt")
        lines (str/split-lines input)]
    (println (str "Part 1: " (part1 lines)))
    (println (str "Part 2: " (part2 lines)))))

(-main)

File: 2025/day10/java/solution.java
import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.regex.*;
import java.math.BigInteger;

public class solution {

    // Part 1: Parse machine line for indicator lights and buttons
    static class MachinePart1 {
        int nLights;
        long target;
        long[] buttonMasks;

        MachinePart1(int nLights, long target, long[] buttonMasks) {
            this.nLights = nLights;
            this.target = target;
            this.buttonMasks = buttonMasks;
        }
    }

    static MachinePart1 parseLinePart1(String line) {
        // Extract indicator pattern [.##.]
        Pattern indicatorPattern = Pattern.compile("\\[([.#]+)\\]");
        Matcher indicatorMatcher = indicatorPattern.matcher(line);
        if (!indicatorMatcher.find()) {
            throw new RuntimeException("No indicator pattern found");
        }

        String indicator = indicatorMatcher.group(1);
        int nLights = indicator.length();

        // Target state: 1 where # appears
        long target = 0;
        for (int i = 0; i < indicator.length(); i++) {
            if (indicator.charAt(i) == '#') {
                target |= (1L << i);
            }
        }

        // Extract button schematics (0,1,2) etc.
        Pattern buttonPattern = Pattern.compile("\\(([0-9,]+)\\)");
        Matcher buttonMatcher = buttonPattern.matcher(line);
        List<Long> buttons = new ArrayList<>();

        while (buttonMatcher.find()) {
            String[] indices = buttonMatcher.group(1).split(",");
            long mask = 0;
            for (String idx : indices) {
                mask |= (1L << Integer.parseInt(idx));
            }
            buttons.add(mask);
        }

        long[] buttonArray = new long[buttons.size()];
        for (int i = 0; i < buttons.size(); i++) {
            buttonArray[i] = buttons.get(i);
        }

        return new MachinePart1(nLights, target, buttonArray);
    }

    // Part 2: Parse machine line for joltage requirements
    static class MachinePart2 {
        int nCounters;
        int[] joltage;
        List<Integer>[] buttons;

        MachinePart2(int nCounters, int[] joltage, List<Integer>[] buttons) {
            this.nCounters = nCounters;
            this.joltage = joltage;
            this.buttons = buttons;
        }
    }

    @SuppressWarnings("unchecked")
    static MachinePart2 parseLinePart2(String line) {
        // Extract joltage requirements {3,5,4,7}
        Pattern joltagePattern = Pattern.compile("\\{([0-9,]+)\\}");
        Matcher joltageMatcher = joltagePattern.matcher(line);
        if (!joltageMatcher.find()) {
            throw new RuntimeException("No joltage pattern found");
        }

        String[] joltageStrs = joltageMatcher.group(1).split(",");
        int[] joltage = new int[joltageStrs.length];
        for (int i = 0; i < joltageStrs.length; i++) {
            joltage[i] = Integer.parseInt(joltageStrs[i]);
        }
        int nCounters = joltage.length;

        // Extract button schematics (0,1,2) etc.
        Pattern buttonPattern = Pattern.compile("\\(([0-9,]+)\\)");
        Matcher buttonMatcher = buttonPattern.matcher(line);
        List<List<Integer>> buttonsList = new ArrayList<>();

        while (buttonMatcher.find()) {
            String[] indices = buttonMatcher.group(1).split(",");
            List<Integer> button = new ArrayList<>();
            for (String idx : indices) {
                button.add(Integer.parseInt(idx));
            }
            buttonsList.add(button);
        }

        List<Integer>[] buttons = new List[buttonsList.size()];
        for (int i = 0; i < buttonsList.size(); i++) {
            buttons[i] = buttonsList.get(i);
        }

        return new MachinePart2(nCounters, joltage, buttons);
    }

    // Count set bits
    static int countBits(long n) {
        return Long.bitCount(n);
    }

    // Solve Part 1: brute force all combinations
    static int solveMachinePart1(MachinePart1 machine) {
        int nButtons = machine.buttonMasks.length;
        int minPresses = Integer.MAX_VALUE;

        // Try all 2^n_buttons combinations
        for (long mask = 0; mask < (1L << nButtons); mask++) {
            long state = 0;
            int presses = 0;

            for (int i = 0; i < nButtons; i++) {
                if ((mask & (1L << i)) != 0) {
                    state ^= machine.buttonMasks[i];
                    presses++;
                }
            }

            if (state == machine.target) {
                minPresses = Math.min(minPresses, presses);
            }
        }

        return minPresses == Integer.MAX_VALUE ? 0 : minPresses;
    }

    // Rational number class for Part 2
    static class Rational {
        BigInteger num;
        BigInteger den;

        Rational(long n) {
            this(BigInteger.valueOf(n), BigInteger.ONE);
        }

        Rational(BigInteger n, BigInteger d) {
            if (d.equals(BigInteger.ZERO)) {
                throw new ArithmeticException("Division by zero");
            }
            if (d.compareTo(BigInteger.ZERO) < 0) {
                n = n.negate();
                d = d.negate();
            }
            BigInteger g = n.gcd(d);
            this.num = n.divide(g);
            this.den = d.divide(g);
        }

        Rational add(Rational other) {
            BigInteger n = this.num.multiply(other.den).add(other.num.multiply(this.den));
            BigInteger d = this.den.multiply(other.den);
            return new Rational(n, d);
        }

        Rational subtract(Rational other) {
            BigInteger n = this.num.multiply(other.den).subtract(other.num.multiply(this.den));
            BigInteger d = this.den.multiply(other.den);
            return new Rational(n, d);
        }

        Rational multiply(Rational other) {
            return new Rational(this.num.multiply(other.num), this.den.multiply(other.den));
        }

        Rational divide(Rational other) {
            return new Rational(this.num.multiply(other.den), this.den.multiply(other.num));
        }

        Rational negate() {
            return new Rational(this.num.negate(), this.den);
        }

        boolean isZero() {
            return num.equals(BigInteger.ZERO);
        }

        boolean isNegative() {
            return num.compareTo(BigInteger.ZERO) < 0;
        }

        boolean isInteger() {
            return den.equals(BigInteger.ONE);
        }

        int toInt() {
            return num.divide(den).intValue();
        }

        double toDouble() {
            return num.doubleValue() / den.doubleValue();
        }

        @Override
        public boolean equals(Object o) {
            if (!(o instanceof Rational)) return false;
            Rational r = (Rational) o;
            return num.equals(r.num) && den.equals(r.den);
        }

        @Override
        public String toString() {
            if (den.equals(BigInteger.ONE)) return num.toString();
            return num + "/" + den;
        }
    }

    // Solve Part 2: Gaussian elimination with rational arithmetic
    static int solveMachinePart2(MachinePart2 machine) {
        int nButtons = machine.buttons.length;
        int nCounters = machine.nCounters;

        if (nButtons == 0) {
            for (int j : machine.joltage) {
                if (j != 0) return Integer.MAX_VALUE;
            }
            return 0;
        }

        // Build matrix A (nCounters x nButtons)
        Rational[][] A = new Rational[nCounters][nButtons];
        for (int i = 0; i < nCounters; i++) {
            for (int j = 0; j < nButtons; j++) {
                A[i][j] = new Rational(0);
            }
        }

        for (int j = 0; j < nButtons; j++) {
            for (int idx : machine.buttons[j]) {
                if (idx < nCounters) {
                    A[idx][j] = new Rational(1);
                }
            }
        }

        Rational[] b = new Rational[nCounters];
        for (int i = 0; i < nCounters; i++) {
            b[i] = new Rational(machine.joltage[i]);
        }

        // Augmented matrix [A | b]
        Rational[][] aug = new Rational[nCounters][nButtons + 1];
        for (int i = 0; i < nCounters; i++) {
            for (int j = 0; j < nButtons; j++) {
                aug[i][j] = A[i][j];
            }
            aug[i][nButtons] = b[i];
        }

        // Gaussian elimination with partial pivoting
        List<int[]> pivotCols = new ArrayList<>();
        int pivotRow = 0;

        for (int col = 0; col < nButtons; col++) {
            // Find non-zero entry in this column
            int found = -1;
            for (int row = pivotRow; row < nCounters; row++) {
                if (!aug[row][col].isZero()) {
                    found = row;
                    break;
                }
            }

            if (found == -1) continue;

            // Swap rows
            Rational[] temp = aug[pivotRow];
            aug[pivotRow] = aug[found];
            aug[found] = temp;

            pivotCols.add(new int[]{col, pivotRow});

            // Scale pivot row
            Rational scale = aug[pivotRow][col];
            for (int c = 0; c <= nButtons; c++) {
                aug[pivotRow][c] = aug[pivotRow][c].divide(scale);
            }

            // Eliminate column in other rows
            for (int row = 0; row < nCounters; row++) {
                if (row != pivotRow && !aug[row][col].isZero()) {
                    Rational factor = aug[row][col];
                    for (int c = 0; c <= nButtons; c++) {
                        aug[row][c] = aug[row][c].subtract(factor.multiply(aug[pivotRow][c]));
                    }
                }
            }

            pivotRow++;
        }

        // Check for inconsistency
        for (int row = pivotRow; row < nCounters; row++) {
            if (!aug[row][nButtons].isZero()) {
                return Integer.MAX_VALUE; // No solution
            }
        }

        // Identify free variables
        Set<Integer> pivotColSet = new HashSet<>();
        for (int[] pc : pivotCols) {
            pivotColSet.add(pc[0]);
        }

        List<Integer> freeVars = new ArrayList<>();
        for (int c = 0; c < nButtons; c++) {
            if (!pivotColSet.contains(c)) {
                freeVars.add(c);
            }
        }

        int nFree = freeVars.size();

        // If no free variables, unique solution
        if (nFree == 0) {
            Rational[] solution = new Rational[nButtons];
            for (int i = 0; i < nButtons; i++) {
                solution[i] = new Rational(0);
            }

            for (int[] pc : pivotCols) {
                int col = pc[0];
                int row = pc[1];
                solution[col] = aug[row][nButtons];
            }

            int total = 0;
            for (Rational val : solution) {
                if (val.isNegative() || !val.isInteger()) {
                    return Integer.MAX_VALUE;
                }
                total += val.toInt();
            }
            return total;
        }

        // Extract null space vectors
        Rational[][] nullVectors = new Rational[nFree][nButtons];
        for (int i = 0; i < nFree; i++) {
            int fv = freeVars.get(i);
            for (int j = 0; j < nButtons; j++) {
                nullVectors[i][j] = new Rational(0);
            }
            nullVectors[i][fv] = new Rational(1);

            for (int[] pc : pivotCols) {
                int col = pc[0];
                int row = pc[1];
                nullVectors[i][col] = aug[row][fv].negate();
            }
        }

        // Extract particular solution
        Rational[] particular = new Rational[nButtons];
        for (int i = 0; i < nButtons; i++) {
            particular[i] = new Rational(0);
        }
        for (int[] pc : pivotCols) {
            int col = pc[0];
            int row = pc[1];
            particular[col] = aug[row][nButtons];
        }

        // Search for optimal solution
        int maxJ = 0;
        for (int j : machine.joltage) {
            maxJ = Math.max(maxJ, j);
        }
        if (maxJ == 0) maxJ = 100;

        return searchOptimalSolution(nButtons, nFree, particular, nullVectors, maxJ);
    }

    static int searchOptimalSolution(int nButtons, int nFree, Rational[] particular, Rational[][] nullVectors, int maxJ) {
        if (nFree == 1) {
            return searchOptimal1D(nButtons, particular, nullVectors[0], maxJ);
        } else if (nFree == 2) {
            return searchOptimal2D(nButtons, particular, nullVectors, maxJ);
        } else if (nFree == 3) {
            return searchOptimal3D(nButtons, particular, nullVectors, maxJ);
        } else if (nFree <= 6) {
            return searchOptimalND(nButtons, nFree, particular, nullVectors, maxJ);
        }
        return 0; // Fallback
    }

    static int searchOptimal1D(int nButtons, Rational[] particular, Rational[] nullVector, int maxJ) {
        double tLow = Double.NEGATIVE_INFINITY;
        double tHigh = Double.POSITIVE_INFINITY;

        for (int j = 0; j < nButtons; j++) {
            double p = particular[j].toDouble();
            double nv = nullVector[j].toDouble();

            if (nv == 0) {
                if (p < 0) return Integer.MAX_VALUE;
            } else if (nv > 0) {
                tLow = Math.max(tLow, -p / nv);
            } else {
                tHigh = Math.min(tHigh, -p / nv);
            }
        }

        if (tLow > tHigh) return Integer.MAX_VALUE;

        int tLowInt = (int) Math.ceil(tLow);
        int tHighInt = (int) Math.floor(tHigh);

        int minTotal = Integer.MAX_VALUE;
        for (int t = tLowInt; t <= tHighInt; t++) {
            Rational tFrac = new Rational(t);
            int total = 0;
            boolean valid = true;

            for (int j = 0; j < nButtons; j++) {
                Rational val = particular[j].add(tFrac.multiply(nullVector[j]));
                if (val.isNegative() || !val.isInteger()) {
                    valid = false;
                    break;
                }
                total += val.toInt();
            }

            if (valid) {
                minTotal = Math.min(minTotal, total);
            }
        }

        return minTotal;
    }

    static int searchOptimal2D(int nButtons, Rational[] particular, Rational[][] nullVectors, int maxJ) {
        int minTotal = Integer.MAX_VALUE;

        int bound = maxJ * 2;
        for (int t0 = -bound; t0 <= bound; t0++) {
            Rational t0Frac = new Rational(t0);
            Rational[] inter0 = new Rational[nButtons];
            for (int j = 0; j < nButtons; j++) {
                inter0[j] = particular[j].add(t0Frac.multiply(nullVectors[0][j]));
            }

            double t1Low = Double.NEGATIVE_INFINITY;
            double t1High = Double.POSITIVE_INFINITY;

            for (int j = 0; j < nButtons; j++) {
                double p = inter0[j].toDouble();
                double nv = nullVectors[1][j].toDouble();

                if (nv > 0) {
                    t1Low = Math.max(t1Low, -p / nv);
                } else if (nv < 0) {
                    t1High = Math.min(t1High, -p / nv);
                }
            }

            int t1LowInt = (int) Math.ceil(t1Low);
            int t1HighInt = (int) Math.floor(t1High);

            for (int t1 = t1LowInt; t1 <= t1HighInt; t1++) {
                Rational t1Frac = new Rational(t1);
                int total = 0;
                boolean valid = true;

                for (int j = 0; j < nButtons; j++) {
                    Rational val = inter0[j].add(t1Frac.multiply(nullVectors[1][j]));
                    if (val.isNegative() || !val.isInteger()) {
                        valid = false;
                        break;
                    }
                    total += val.toInt();
                }

                if (valid && total < minTotal) {
                    minTotal = total;
                }
            }
        }

        return minTotal;
    }

    static int searchOptimal3D(int nButtons, Rational[] particular, Rational[][] nullVectors, int maxJ) {
        int minTotal = Integer.MAX_VALUE;
        int bound = maxJ;

        for (int t0 = -bound; t0 <= bound; t0++) {
            Rational t0Frac = new Rational(t0);
            Rational[] inter0 = new Rational[nButtons];
            for (int j = 0; j < nButtons; j++) {
                inter0[j] = particular[j].add(t0Frac.multiply(nullVectors[0][j]));
            }

            double t1Low = Double.NEGATIVE_INFINITY;
            double t1High = Double.POSITIVE_INFINITY;

            for (int j = 0; j < nButtons; j++) {
                double p = inter0[j].toDouble();
                double nv = nullVectors[1][j].toDouble();

                if (nv > 0) {
                    t1Low = Math.max(t1Low, -p / nv - bound);
                } else if (nv < 0) {
                    t1High = Math.min(t1High, -p / nv + bound);
                }
            }

            int t1LowInt = Math.max((int) Math.ceil(t1Low), -bound);
            int t1HighInt = Math.min((int) Math.floor(t1High), bound);

            for (int t1 = t1LowInt; t1 <= t1HighInt; t1++) {
                Rational t1Frac = new Rational(t1);
                Rational[] inter1 = new Rational[nButtons];
                for (int j = 0; j < nButtons; j++) {
                    inter1[j] = inter0[j].add(t1Frac.multiply(nullVectors[1][j]));
                }

                double t2Low = Double.NEGATIVE_INFINITY;
                double t2High = Double.POSITIVE_INFINITY;

                for (int j = 0; j < nButtons; j++) {
                    double p = inter1[j].toDouble();
                    double nv = nullVectors[2][j].toDouble();

                    if (nv > 0) {
                        t2Low = Math.max(t2Low, -p / nv);
                    } else if (nv < 0) {
                        t2High = Math.min(t2High, -p / nv);
                    }
                }

                int t2LowInt = (int) Math.ceil(t2Low);
                int t2HighInt = (int) Math.floor(t2High);

                for (int t2 = t2LowInt; t2 <= t2HighInt; t2++) {
                    Rational t2Frac = new Rational(t2);
                    int total = 0;
                    boolean valid = true;

                    for (int j = 0; j < nButtons; j++) {
                        Rational val = inter1[j].add(t2Frac.multiply(nullVectors[2][j]));
                        if (val.isNegative() || !val.isInteger()) {
                            valid = false;
                            break;
                        }
                        total += val.toInt();
                    }

                    if (valid && total < minTotal) {
                        minTotal = total;
                    }
                }
            }
        }

        return minTotal;
    }

    static int searchOptimalND(int nButtons, int nFree, Rational[] particular, Rational[][] nullVectors, int maxJ) {
        int[] minTotal = {Integer.MAX_VALUE};
        searchRecursive(0, nFree, nButtons, particular.clone(), nullVectors, maxJ, minTotal);
        return minTotal[0];
    }

    static void searchRecursive(int idx, int nFree, int nButtons, Rational[] partial, Rational[][] nullVectors, int maxJ, int[] minTotal) {
        if (idx == nFree) {
            int total = 0;
            boolean valid = true;
            for (Rational val : partial) {
                if (val.isNegative() || !val.isInteger()) {
                    valid = false;
                    break;
                }
                total += val.toInt();
            }
            if (valid) {
                minTotal[0] = Math.min(minTotal[0], total);
            }
            return;
        }

        double tLow = Double.NEGATIVE_INFINITY;
        double tHigh = Double.POSITIVE_INFINITY;

        for (int j = 0; j < nButtons; j++) {
            double p = partial[j].toDouble();
            double nv = nullVectors[idx][j].toDouble();

            if (nv > 0) {
                tLow = Math.max(tLow, -p / nv);
            } else if (nv < 0) {
                tHigh = Math.min(tHigh, -p / nv);
            }
        }

        if (tLow > tHigh || Double.isInfinite(tLow) || Double.isInfinite(tHigh)) {
            return;
        }

        int tLowInt = Math.max((int) Math.ceil(tLow) - maxJ, -maxJ * 2);
        int tHighInt = Math.min((int) Math.floor(tHigh) + maxJ, maxJ * 2);

        for (int t = tLowInt; t <= tHighInt; t++) {
            Rational tFrac = new Rational(t);
            Rational[] newPartial = new Rational[nButtons];
            for (int j = 0; j < nButtons; j++) {
                newPartial[j] = partial[j].add(tFrac.multiply(nullVectors[idx][j]));
            }
            searchRecursive(idx + 1, nFree, nButtons, newPartial, nullVectors, maxJ, minTotal);
        }
    }

    static int part1(List<String> lines) {
        int total = 0;
        for (String line : lines) {
            line = line.trim();
            if (line.isEmpty()) continue;

            MachinePart1 machine = parseLinePart1(line);
            int minPresses = solveMachinePart1(machine);
            total += minPresses;
        }
        return total;
    }

    static int part2(List<String> lines) {
        int total = 0;
        for (String line : lines) {
            line = line.trim();
            if (line.isEmpty()) continue;

            MachinePart2 machine = parseLinePart2(line);
            int minPresses = solveMachinePart2(machine);
            total += minPresses;
        }
        return total;
    }

    public static void main(String[] args) throws IOException {
        Path inputPath = Paths.get(System.getProperty("user.dir")).getParent().resolve("input.txt");
        List<String> lines = Files.readAllLines(inputPath);

        System.out.println("Part 1: " + part1(lines));
        System.out.println("Part 2: " + part2(lines));
    }
}
